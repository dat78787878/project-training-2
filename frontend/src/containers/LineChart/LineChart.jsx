import React from 'react';
import Chart from 'react-apexcharts';
import { useSelector, useDispatch } from 'react-redux';
import { useCallback, useEffect, useMemo, useState } from 'react';
import moment from 'moment';
import Loading from '../../components/Loading/Loading';
import Error from '../../components/Error/Error';
import { getData } from '../../redux/lineChart/actions';

const LineChart = () => {
  const dispatch = useDispatch();
  const { lineData, isLoading, isError } = useSelector((state) => state.lineChart);
  const { fromDate, toDate } = useSelector((state) => state.dateRange);
  const groupData = ['Day', 'Week', 'Month'];
  const [isActive, setIsActive] = useState(0);
  const [series, setSeries] = useState([]);

  useEffect(() => {
    dispatch(getData([fromDate, toDate]));
  }, [fromDate, toDate]);

  useEffect(() => {
    setSeries(lineData);
    setIsActive(0);
  }, [lineData]);

  const totalData = useMemo(() => {
    const total = lineData.map((val) => val.data.reduce((acc, val) => acc + val.y, 0));
    return total;
  }, [lineData]);

  const options = {
    chart: {
      toolbar: {
        show: false
      }
    },
    legend: {
      show: true,
      horizontalAlign: 'left',
      formatter: (seriesName, opts) => seriesName + ' ' + totalData[opts.seriesIndex]
    },
    colors: ['#00ffc8', '#ea00ff'],
    title: {
      text: 'Decive',
      align: 'left',
      style: {
        fontSize: '18px'
      }
    }
  };

  const handleData = (lineData, formatType) => {
    const newSeries = lineData.map((device) => {
      const data = device.data.map((date) => ({
        x: moment(date.x, 'YYYY-MM-DD').format(formatType),
        y: date.y
      }));

      let obj = {};
      const newData = data.reduce((totalY, date) => {
        if (!obj[date.x]) {
          obj[date.x] = date;
          totalY.push(obj[date.x]);
        } else {
          obj[date.x].y += date.y;
        }
        return totalY;
      }, []);
      return {
        name: device.name,
        data: newData
      };
    });

    return newSeries;
  };

  const handleClick = useCallback(
    (val) => {
      setIsActive(val);
      if (val === 'Day') {
        setSeries(lineData);
      }
      if (val === 'Week') {
        const newSeries = handleData(lineData, 'WW');
        setSeries(newSeries);
      }
      if (val === 'Month') {
        const newSeries = handleData(lineData, 'MM-YYYY');
        setSeries(newSeries);
      }
    },
    [lineData]
  );

  const lineChart = (
    <div className="lineChart-container">
      <div className="lineChart-container-menu">
        {groupData.map((val, index) => {
          return (
            <>
              <span
                key={index}
                style={{ opacity: isActive === val ? 1 : 0.3 }}
                data-testid={val}
                onClick={() => handleClick(val)}>
                <div
                  className={
                    isActive === val
                      ? 'lineChart-container-menu-item'
                      : 'lineChart-container-menu-item none'
                  }></div>
                {val}
              </span>
            </>
          );
        })}
      </div>
      <Chart options={options} series={series} type="line" height={620} />
    </div>
  );
  return (
    <div className="lineChart">
      {isLoading && <Loading />}
      {isError && <Error />}
      {!isLoading && !isError && lineChart}
    </div>
  );
};

export default LineChart;
