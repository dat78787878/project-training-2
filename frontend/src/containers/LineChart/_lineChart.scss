.lineChart {
  background-color: #ffffff;
  height: 485px;
  
  @include pc {
    height: 500px;
  }
  @include tablet {
    height: 495px;
  }
  @include mobile {
    height: 503px;
  }
  &-container {
    &-menu {
      display: flex;
      justify-content: flex-end;
      &-item {
        width: 5px;
        height: 20px;
        margin: 0 auto;
        background-color: #00ff22;
      }

      span {
        font-size: 16px;
        color: #000000;
        font-weight: bold;
        padding-right: 15px;
      }
    }
  }
  .pl{
    padding-left: 35px;
    @include tablet {
      padding-top: 20px;
      margin-top: 10px;
    }
    @include mobile {
      padding-top: 0px;
      margin-top: 10px;
      padding-left: 0px;
    }
    @include laptop {
      margin-top: 10px;
    }
  }
}

.none {
  visibility: hidden;
}
