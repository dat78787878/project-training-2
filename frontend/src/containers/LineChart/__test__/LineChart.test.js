import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import LineChart from '../LineChart';
window.React = React;

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));
jest.mock('react-apexcharts', () => 'Chart');
describe('LineChart component', () => {
  let mockAppState = {};
  beforeEach(() => {
    useDispatch.mockImplementation(() => () => {});
    useSelector.mockImplementation((callback) => {
      return callback(mockAppState);
    });
  });
  afterEach(() => {
    useSelector.mockClear();
    useDispatch.mockClear();
  });
  it('check loading', () => {
    mockAppState = {
      lineChart: {
        lineData: [],
        isLoading: true,
        isError: false
      },
      dateRange: {
        fromDate: '08/08/2021',
        toDate: '09/09/2021'
      }
    };
    const { container } = render(<LineChart />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('div').firstChild.className).toEqual('loading');
  });
  it('success task', () => {
    mockAppState = {
      lineChart: {
        lineData: [
          {
            name: 'Android',
            data: [
              { x: '01/01/2021', y: 4 },
              { x: '02/01/2021', y: 3 },
              { x: '03/01/2021', y: 5 }
            ]
          },
          {
            name: 'iOS',
            data: [
              { x: '01/01/2021', y: 1 },
              { x: '02/01/2021', y: 6 },
              { x: '03/01/2021', y: 4 }
            ]
          }
        ],
        isloading: false,
        isError: false
      },
      dateRange: {
        fromDate: '08/08/2021',
        toDate: '09/09/2021'
      }
    };
    const { container } = render(<LineChart />);
    expect(container).toMatchSnapshot();
    expect(container.getElementsByClassName('apexcharts-canvas'));
  });
  it('get data day', () => {
    mockAppState = {
      lineChart: {
        lineData: [
          {
            name: 'Android',
            data: [
              { x: '01/01/2021', y: 5 },
              { x: '02/01/2021', y: 10 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 4 },
              { x: '02/02/2021', y: 6 }
            ]
          },
          {
            name: 'iOS',
            data: [
              { x: '01/01/2021', y: 2 },
              { x: '02/01/2021', y: 3 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 10 },
              { x: '02/02/2021', y: 5 }
            ]
          }
        ],
        isloading: false,
        isError: false
      },
      dateRange: {
        fromDate: '08/08/2021',
        toDate: '09/09/2021'
      }
    };
    const { container, getByTestId } = render(<LineChart />);
    fireEvent.click(getByTestId('Day'));
    expect(container).toMatchSnapshot();
  });
  it('get data week', () => {
    mockAppState = {
      lineChart: {
        lineData: [
          {
            name: 'Android',
            data: [
              { x: '01/01/2021', y: 5 },
              { x: '02/01/2021', y: 10 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 4 },
              { x: '02/02/2021', y: 6 }
            ]
          },
          {
            name: 'iOS',
            data: [
              { x: '01/01/2021', y: 2 },
              { x: '02/01/2021', y: 3 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 10 },
              { x: '02/02/2021', y: 5 }
            ]
          }
        ],
        isloading: false,
        isError: false
      },
      dateRange: {
        fromDate: '2021-08-08',
        toDate: '2021-08-09'
      }
    };
    const { container, getByTestId } = render(<LineChart />);
    fireEvent.click(getByTestId('Week'));
    expect(container).toMatchSnapshot();
  });
  it('get data month', () => {
    mockAppState = {
      lineChart: {
        lineData: [
          {
            name: 'Android',
            data: [
              { x: '01/01/2021', y: 5 },
              { x: '02/01/2021', y: 10 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 4 },
              { x: '02/02/2021', y: 6 }
            ]
          },
          {
            name: 'iOS',
            data: [
              { x: '01/01/2021', y: 2 },
              { x: '02/01/2021', y: 3 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 10 },
              { x: '02/02/2021', y: 5 }
            ]
          }
        ],
        isloading: false,
        isError: false
      },
      dateRange: {
        fromDate: '2021-08-08',
        toDate: '2021-08-09'
      }
    };
    const { container, getByTestId } = render(<LineChart />);
    fireEvent.click(getByTestId('Month'));
    expect(container).toMatchSnapshot();
  });
  it('error task', () => {
    mockAppState = {
      lineChart: {
        lineData: [],
        isloading: false,
        isError: true
      },
      dateRange: {
        fromDate: '2021-08-08',
        toDate: '2021-08-09'
      }
    };
    const { container, getByText } = render(<LineChart />);
    expect(container).toMatchSnapshot();
    expect(getByText('Error')).toBeTruthy();
  });
});
