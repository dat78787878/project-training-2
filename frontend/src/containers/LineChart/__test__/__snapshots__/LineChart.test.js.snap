// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart component check loading 1`] = `
<div>
  <div
    class="lineChart"
  >
    <div
      class="loading"
    >
      <div
        class="spinner-border text-success"
      />
    </div>
  </div>
</div>
`;

exports[`LineChart component error task 1`] = `
<div>
  <div
    class="lineChart"
  >
    <div
      class="error"
    >
      Error
      <img
        alt="not-found"
        src="https://www.pngitem.com/pimgs/m/561-5616833_image-not-found-png-not-found-404-png.png"
      />
    </div>
  </div>
</div>
`;

exports[`LineChart component get data day 1`] = `
<div>
  <div
    class="lineChart"
  >
    <div
      class="lineChart-container"
    >
      <div
        class="lineChart-container-menu"
      >
        <span
          data-testid="Day"
          style="opacity: 1;"
        >
          <div
            class="lineChart-container-menu-item"
          />
          Day
        </span>
        <span
          data-testid="Week"
          style="opacity: 0.3;"
        >
          <div
            class="lineChart-container-menu-item none"
          />
          Week
        </span>
        <span
          data-testid="Month"
          style="opacity: 0.3;"
        >
          <div
            class="lineChart-container-menu-item none"
          />
          Month
        </span>
      </div>
      <chart
        height="620"
        options="[object Object]"
        series="[object Object],[object Object]"
        type="line"
      />
    </div>
  </div>
</div>
`;

exports[`LineChart component get data month 1`] = `
<div>
  <div
    class="lineChart"
  >
    <div
      class="lineChart-container"
    >
      <div
        class="lineChart-container-menu"
      >
        <span
          data-testid="Day"
          style="opacity: 0.3;"
        >
          <div
            class="lineChart-container-menu-item none"
          />
          Day
        </span>
        <span
          data-testid="Week"
          style="opacity: 0.3;"
        >
          <div
            class="lineChart-container-menu-item none"
          />
          Week
        </span>
        <span
          data-testid="Month"
          style="opacity: 1;"
        >
          <div
            class="lineChart-container-menu-item"
          />
          Month
        </span>
      </div>
      <chart
        height="620"
        options="[object Object]"
        series="[object Object],[object Object]"
        type="line"
      />
    </div>
  </div>
</div>
`;

exports[`LineChart component get data week 1`] = `
<div>
  <div
    class="lineChart"
  >
    <div
      class="lineChart-container"
    >
      <div
        class="lineChart-container-menu"
      >
        <span
          data-testid="Day"
          style="opacity: 0.3;"
        >
          <div
            class="lineChart-container-menu-item none"
          />
          Day
        </span>
        <span
          data-testid="Week"
          style="opacity: 1;"
        >
          <div
            class="lineChart-container-menu-item"
          />
          Week
        </span>
        <span
          data-testid="Month"
          style="opacity: 0.3;"
        >
          <div
            class="lineChart-container-menu-item none"
          />
          Month
        </span>
      </div>
      <chart
        height="620"
        options="[object Object]"
        series="[object Object],[object Object]"
        type="line"
      />
    </div>
  </div>
</div>
`;

exports[`LineChart component success task 1`] = `
<div>
  <div
    class="lineChart"
  >
    <div
      class="lineChart-container"
    >
      <div
        class="lineChart-container-menu"
      >
        <span
          data-testid="Day"
          style="opacity: 0.3;"
        >
          <div
            class="lineChart-container-menu-item none"
          />
          Day
        </span>
        <span
          data-testid="Week"
          style="opacity: 0.3;"
        >
          <div
            class="lineChart-container-menu-item none"
          />
          Week
        </span>
        <span
          data-testid="Month"
          style="opacity: 0.3;"
        >
          <div
            class="lineChart-container-menu-item none"
          />
          Month
        </span>
      </div>
      <chart
        height="620"
        options="[object Object]"
        series="[object Object],[object Object]"
        type="line"
      />
    </div>
  </div>
</div>
`;
